python/tank/util/shotgun.py

```diff
@@ -166,13 +166,68 @@ def __get_sg_config_data(shotgun_cfg_path, user="default"):
     # validate the config data to ensure all fields are present
     if "host" not in config_data:
         raise TankError("Missing required field 'host' in config '%s'" % shotgun_cfg_path)
-    if "api_script" not in config_data:
-        raise TankError("Missing required field 'api_script' in config '%s'" % shotgun_cfg_path)
-    if "api_key" not in config_data:
-        raise TankError("Missing required field 'api_key' in config '%s'" % shotgun_cfg_path)
-    
+
     return config_data
 
+
+def _is_script_user_authenticated(config_data):
+    """
+    Indicates if we are authenticating with a script user for a given configuration.
+    :param config_data: The configuration data
+    :returns: True is we are using a script user, False otherwise.
+    """
+    return config_data.get("api_script") and config_data.get("api_key")
+
+
+def is_script_user_authenticated():
+    """
+    Indicates if we are authenticating with a script user this core's configuration.
+    :returns: True is we are using a script user, False otherwise.
+    """
+    return _is_script_user_authenticated(get_associated_sg_config_data())
+
+
+def is_authenticated():
+    """
+    Indicates if we need to authenticate.
+    :returns: True is we are using a script user or have a valid session token, False
+    """
+    # FIXME: Break circular dependency this way, not really clean...
+    from . import session
+    config_data = get_associated_sg_config_data()
+    return _is_script_user_authenticated(config_data) or\
+        session.create_sg_connection_from_session(config_data)
+
+
+def __create_session_based_sg_connection(config_data):
+    """
+    Creates a shotgun connection using the current session token or a new one if the old one
+    expired.
+    :param config_data: A dictionary holding the "host" and "http_proxy"
+    :returns: A valid Shotgun instance.
+    :raises TankAuthenticationError: If we couldn't get a valid session, a TankError is thrown.
+    """
+    # FIXME: Break circular dependency this way, not really clean...
+    from ..platform import current_engine
+    from . import session
+
+    # If the Shotgun login was not automated, then try to create a Shotgun
+    # instance from the cached session id.
+    sg = session.create_sg_connection_from_session(config_data)
+    # If that didn't work
+    if not sg:
+        # If there is a current engine, we can ask the engine to prompt the user to login
+        if current_engine():
+            current_engine().renew_session()
+            sg = session.create_sg_connection_from_session(config_data)
+        else:
+            # Otherwise we failed and can't login.
+            raise TankAuthenticationError("No authentification credentials were found.")
+    if not sg:
+        raise TankAuthenticationError("Authentification failed.")
```

Comment:
Yup, you are totally right. I caught this mistake while writing the unit test as well. This was a late add that hadn’t been retested manually.